lib/pages-router.js

```diff
@@ -0,0 +1,162 @@
+            }
+        }
+        return (appData.context + "/");
+    }
+
+    /**
+     * Whether the specified page is processable or not.
+     * @param page {UIComponent} page
+     * @param renderingContext {RenderingContext} rendering context
+     * @param lookupTable {LookupTable} lookup table
+     * @param request {Object} HTTP request
+     * @param response {Object} HTTP response
+     * @return {boolean} <code>true</code> if processable, otherwise <code>false</code>
+     */
+    function isPageProcessable(page, renderingContext, lookupTable, request, response) {
+        if (page.disabled) {
+            // This page is disabled.
+            response.sendError(404, "Requested page '" + renderingContext.pageUri
+                                    + "' does not exists.");
+            return false;
+        }
+
+        var pageDefinition = page.definition;
+        if (utils.parseBoolean(pageDefinition[constants.UI_COMPONENT_DEFINITION_IS_ANONYMOUS])) {
+            // This is an anonymous page. So no need for an user session or checking permissions.
+            return true;
+        }
+
+        // This is not an anonymous page.
+        var user = renderingContext.user;
+        if (user) {
+            // An user has logged in.
+            var pagePermissions = pageDefinition[constants.UI_COMPONENT_DEFINITION_PERMISSIONS];
+            if (pagePermissions && Array.isArray(pagePermissions)) {
+                // A permissions array is specified in the page definition.
+                var numberOfUnitPermissions = pagePermissions.length;
+                var userPermissionsMap = user.permissions;
+                for (var i = 0; i < numberOfUnitPermissions; i++) {
+                    if (!userPermissionsMap.hasOwnProperty(pagePermissions[i])) {
+                        // User does not have this permission.
+                        if (log.isDebugEnabled()) {
+                            log.debug("User '" + user.username + "' in domain '" + user.domain
+                                      + "' does not have permission '" + pagePermissions[i]
+                                      + "' to view page '" + page.fullName + "'.");
+                        }
+                        response.sendError(403, "You do not have enough permissions to access the "
+                                                + "requested page '" + renderingContext.pageUri
+                                                + "'.");
+                        return false;
+                    }
+                }
+                // User has all permissions.
+                return true;
+            } else {
+                // Permissions are not specified in the page definition.
+                return true;
+            }
+        } else {
+            // Currently no user has logged in. So redirect to the login page.
+            var redirectUri = getLoginPageUri(renderingContext, lookupTable) + "?"
```

Comment:
vulnerable for redirect attacks https://www.owasp.org/index.php/Unvalidated_Redirects_and_Forwards_Cheat_Sheet